classes/patreon_login.php

```diff
@@ -14,72 +14,208 @@ public static function updateExistingUser($user_id, $user_response, $tokens) {
 
 		$user = wp_get_current_user();
 
 		if(0 == $user->ID) {
+			
+			$redirect = add_query_arg( 'patreon_message', 'patreon_cant_login_strict_oauth', $redirect);
+			wp_redirect($redirect);
+			exit;			
 		} else {
 			/* update user meta data with patreon data */
 			self::updateExistingUser($user->ID, $user_response, $tokens);
+			wp_redirect($redirect);
+			exit;				
 		}
 
 		if($redirect == false || is_null($redirect) ) {
 			wp_redirect(home_url());
 			exit;
 		} else {
-			wp_redirect( get_permalink($redirect) );
+			wp_redirect( $redirect );
 			exit;
 		}
 
 	}
 
-	public static function createUserFromPatreon($user_response, $tokens, $redirect = false) {
+	public static function checkTokenExpiration($user_id=false) {
+
+		if($user_id)
+		{
+			$user = get_user_by('ID',$user_id);
+		}
+		else
+		{
+			$user = wp_get_current_user();
+		}
+
+		if(!$user OR 0 == $user->ID) {
+		} else {
+			
+			// Valid user is logged in. Check the token:
+			
+			$expiration = get_user_meta($user->ID,'patreon_token_expires_in',true);
+			$minted = get_user_meta($user->ID,'patreon_token_minted',true);
+			
+			if($minted!='')
+			{
+				// We have value. get secs to use them in comparison.
+
+				$minted = explode(' ',$minted);
+				// Cast to integer
+				$minted = (int) $minted[1];
+				
+				if((int)microtime(true) >= ($minted+$expiration))
+				{
+					// This token is expired. Nuke it.
+					
+					delete_user_meta($user->ID,'patreon_access_token');
+			
+				}
+		
+			}
+			else
+			{
+				// No minted value. Even if there may be no access token created and saved, still nuke it.
+				
+				delete_user_meta($user->ID,'patreon_access_token');
+				
+			}
+	
+		}
+
+
+	}
+
+	public static function createOrLogInUserFromPatreon($user_response, $tokens, $redirect = false) {
 
 		global $wpdb;
 
-		$login_with_patreon = get_option('patreon-enable-login-with-patreon', false);
+		$login_with_patreon = get_option('patreon-enable-login-with-patreon', true);
+
 		$admins_editors_login_with_patreon = get_option('patreon-enable-allow-admins-login-with-patreon', false);
+		
+		$danger_user_list = Patreon_Login::getDangerUserList();
+		
+		// Chec if user is logged in to wp:
+		
+		// Logged in user. We just link him up and be done.
+		if(is_user_logged_in()) {
+			
+			$user = wp_get_current_user();
+						
+			self::updateExistingUser($user->ID, $user_response, $tokens);
+			wp_redirect($redirect);
+			exit;		
 
-		$email = $user_response['data']['attributes']['email'];
+		}
+		
+		////////////////////////////////////////////////
+		// If we are here, User is not logged in. Go through login or creation procedure:
+		////////////////////////////////////////////////
+		
+		$patreon_user_id = $user_response['data']['id'];
+		
+		// First see if any user was linked to this patreon user:
+		
+		global $wpdb;
+		
+		$prepared_sql = $wpdb->prepare(
+		
+						"SELECT * FROM ".$wpdb->usermeta." WHERE meta_key = 'patreon_user_id' AND meta_value = %s",
```

Comment:
Is this another case of weird github formatting?

Reply:
Nope, they seemed to be indented 4 times more than they should have

https://github.com/codebard/patreon-wordpress/commit/84d2ae3d1103f462b07be0cbfd7726370a3366b2